[KOGITO-4322] Add startup probes

Kevin-Mok opened the pull request:
See: https://issues.redhat.com/browse/KOGITO-4322

Downstream PR: https://github.com/kiegroup/rhpam-kogito-operator/pull/12

## Description
This PR allows the operator to add startup probes to the deployment if the Kubernetes cluster version is ≥1.18 when the feature is enabled by default.

## Testing
I deployed with the following YAML:
```yaml
apiVersion: app.kiegroup.org/v1beta1
kind: KogitoRuntime
metadata:
  name: qe
spec:
  replicas: 1
  image: quay.io/kmok/process-quarkus-example:health-no-redirect
  probes:
    startupProbe:
      failureThreshold: 60
      periodSeconds: 1
```
Where I set the `periodSeconds` to the minimum amount such that it was bound to fail and show up in the pod events:
```
Warning  Unhealthy 28s (x5 over 32s)  kubelet, crc-j55b9-master-0  Startup probe failed: Get http://10.116.0.168:8080/q/health/live: dial tcp 10.116.0.168:8080: connect: connection refused 
```
I assume it to be working as such.

## Requirements
Many thanks for submiting your Pull Request :heart:! 

Please make sure that your PR meets the following requirements:

- [x] You have read the [contributors' guide](https://github.com/kiegroup/kogito-cloud-operator/blob/master/README.md#contributing-to-the-kogito-operator)
- [x] Pull Request title is properly formatted: 
- [x] Pull Request contains link to the JIRA issue
- [x] Pull Request contains description of the issue
- [x] Pull Request does not include fixes for issues other than the main ticket
- [x] Your feature/bug fix has a unit test that verifies it
- [x] You've tested the new feature/bug fix in an actual OpenShift cluster
- [x] You've added a [RELEASE_NOTES.md](RELEASE_NOTES.md) entry regarding this change

Kevin-Mok commented:
/jenkins test

ricardozanini commented on api/probe.go:
```diff
@@ -22,4 +22,6 @@ type KogitoProbeInterface interface {
 	SetLivenessProbe(livenessProbe corev1.Probe)
 	GetReadinessProbe() corev1.Probe
 	SetReadinessProbe(readinessProbe corev1.Probe)
+	GetStartupProbe() corev1.Probe
+	SetStartupProbe(startupProbe corev1.Probe)
```
@Kevin-Mok can you please also open a PR to downstream operator and implement the interface there? Since we are updating the interface, we also need to update product CRDs. Thanks!

ricardozanini replied:
README? what README, @spolti ? Maybe the docs? Yes, the docs would be a good fit.

spolti replied:
@ricardozanini sry, I meant about the need to replicate the interface methods, not about the feature itself..

spolti replied:
@Kevin-Mok did you create the PR to update the rhpam-kogito-operator?

Kevin-Mok replied:
I just opened the [downstream PR](https://github.com/kiegroup/rhpam-kogito-operator/pull/12).

ricardozanini commented on core/kogitoservice/deployment.go:
```diff
@@ -99,10 +101,26 @@ func (d *deploymentHandler) CreateRequiredDeployment(service api.KogitoService,
+	minorVersion := strings.TrimSuffix(versionInfo.Minor, "+")
+	if minorVersionInt, err := strconv.Atoi(minorVersion); err != nil {
+		d.Log.Warn("Could not parse Kubernetes server minor version. Startup probes will not be added.")
+	} else if minorVersionInt >= 18 {
```
```suggestion
	} else if minorVersionInt >= 18 {
```

Can you please add `18` to a constant?

ricardozanini reviewed (changes requested):
Great work! Small nitpicks only! Many thanks, man!

spolti commented on core/kogitoservice/probe.go:
```diff
@@ -79,7 +90,12 @@ func getQuarkusHealthCheckLiveness(probe corev1.Probe) *corev1.Probe {
 	}
+	setDefaultProbeValues(&probe)
+	// Must be 1 for liveness and startup.
+	probe.SuccessThreshold = 1
```
wouldn't be better to enforce it when it can be different than 1?

Kevin-Mok replied:
Right. So with this line, all liveness/startup probes returned will have a `SuccessThreshold` value of 1.

spolti replied:
@ricardozanini sry, I meant about the need to replicate the interface methods, not about the feature itself.

sutaakar commented:
The code looks good, will do a full review once https://github.com/kiegroup/kogito-operator/pull/822 is merged.

spolti reviewed (approved):
Approving, before merge let's wait for the needed changes on rhpam-kogito-operator.

Kevin-Mok commented:
@spolti This PR is rebased on top of [KOGITO-4781](https://github.com/kiegroup/kogito-operator/pull/822) since that fixes the issue that this PR triggers even more, so it should be merged after it. @sutaakar will also do his full review after KOGITO-4781 is merged.

spolti commented:
looks it needs a rebase.

Do this change depends on https://github.com/kiegroup/kogito-operator/pull/822 ?

Kevin-Mok commented:
@spolti Yes, it does. I will rebase after #822 is merged.

Kevin-Mok commented:
/jenkins test

sutaakar reviewed (approved):
Looks good to me, works as expected

sutaakar reviewed (changes requested):
I hit an endless reconciliation with:
```

apiVersion: app.kiegroup.org/v1beta1
kind: KogitoRuntime
metadata:
  name: example-springboot
spec:
  # see the quarkus-jvm.Dockerfile in this directory to know how to build your image locally
  image: quay.io/kiegroup/examples-process-springboot-example-nightly:latest
  runtime: springboot
```

sutaakar commented:
Seems to be issue specific for TCP sockets and startup probe:
```
{"level":"info","ts":1618312624.1972435,"logger":"comparator","msg":"Objects are not equal",

"deployed":......"livenessProbe":{"tcpSocket":{"port":8080}},"readinessProbe":{"tcpSocket":{"port":8080}},"startupProbe":{"tcpSocket":{"port":8080},"timeoutSeconds":1,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},.....

"requested":....,"livenessProbe":{"tcpSocket":{"port":8080}},"readinessProbe":{"tcpSocket":{"port":8080}},"startupProbe":{"tcpSocket":{"port":8080}},"imagePullPolicy":"Always"}],....
```

ricardozanini commented:
^ @Kevin-Mok you can either customize the comparator for the probes (see the DeploymentComparator) or you can set the defaults. I think the former is a better option.

Kevin-Mok commented:
/jenkins test

@sutaakar I initially forgot to add the function call to `setDefaultProbeValues` in `getTCPHealthCheckProbe`, so in this PR, it's missing the line. But in the PR for KOGITO-4781, I added the line, rebased this PR on top of it, so it should be working now.

sutaakar reviewed (approved):
Works as expected now, thanks.

spolti commented:
is it ready to merge?

Kevin-Mok commented:
I'm pretty sure this is ready to merge, but I'll ping @ricardozanini here just to make sure.
